### PR TITLE
Convert macros to rzshell

### LINKED ISSUE #1584: Port `(` (macroses) to rzshell

### PR DESCRIPTION
**Pull request by ret2libc**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Convert the existing macros functionality to rzshell. The grammar parses the various tokens that compose a macro statement and we use a completely separated handler to handle macros than the others command handlers, due to the different kind of arguments it takes.

`(`(to list macros) and `(-`(to remove macros) are still handled as regular `arged_stmt`. I could not make `(` output json/rizin cmds or other kind of outputs because `(j` would be hard to differentiate between `(`JSON mode or a wrongly-written command to add a new macro named `j`.

Bonus: macros arguments are replaced with their names and not with numbers anymore.
Before:
```
(foo a b; pd $0 @ $1)
```
After:
```
(foo a b; pd ${a} @ ${b})
```

**Test plan**

Green tests. Unit tests for the new macros API added.

**Closing issues**

Closes #1584

### CODE REVIEW

**Comment by ret2libc** [comment]
Maybe to avoid problems it would be better to use `${xy}` instead of just `$xy`(which would be problematic when you want to add some strings after the replacement (e.g. `$xyworld` => `${xy}world`).

**Comment by XVilka** [comment]
@ret2libc, yes, makes sense, especially since we already use `$` aliases.

File: librz/core/cmd/cmd.c [hunk 1]
```diff
@@ -3719,6 +3672,52 @@ DEFINE_HANDLE_TS_FCN_AND_SYMBOL(arged_stmt) {
...
+
+	TSNode macro_name = ts_node_child_by_field_name(macro_content, "name", strlen("name"));
+	rz_return_val_if_fail(!ts_node_is_null(macro_name), false);
+	macro_name_str = ts_node_sub_string(macro_name, state->input);
+
+	TSNode args = ts_node_child_by_field_name(macro_content, "args", strlen("args"));
+	if (!ts_node_is_null(args)) {
+		nargs = ts_node_named_child_count(args);
+	}
+	args_str = RZ_NEWS0(char *, nargs + 1);
```
Comment (XVilka) [inline librz/core/cmd/cmd.c:3696]: Check for NULL

File: librz/core/cmd/cmd.c [hunk 2]
```diff
@@ -3719,6 +3672,52 @@ DEFINE_HANDLE_TS_FCN_AND_SYMBOL(arged_stmt) {
...
+
+	TSNode body = ts_node_child_by_field_name(macro_content, "body", strlen("body"));
+	rz_return_val_if_fail(!ts_node_is_null(body), false);
+	body_str = ts_node_sub_string(body, state->input);
+
+	TSNode macro_call = ts_node_named_child(node, 2);
+	if (!ts_node_is_null(macro_call)) {
+		TSNode argv = ts_node_child_by_field_name(macro_call, "argv", strlen("argv"));
+		ut32 nargv = ts_node_is_null(argv) ? 0 : ts_node_named_child_count(argv);
+		argv_str = RZ_NEWS0(char *, nargv + 1);
```
Comment (XVilka) [inline librz/core/cmd/cmd.c:3716]: Check for NULL

File: librz/core/cmd/cmd.c [hunk 3]
```diff
@@ -5217,8 +5214,8 @@ struct ts_data_symbol_map map_ts_symbols[] = {
 /**
  * \brief Create an instance of RzCmd for the Rizin language
  */
-RZ_API RzCmd *rz_core_cmd_new(bool has_cons) {
-	RzCmd *res = rz_cmd_new(has_cons);
+RZ_API RzCmd *rz_core_cmd_new(RzCore *core, bool has_cons) {
```
Comment (XVilka) [inline librz/core/cmd/cmd.c:5237]: It would require updates of the plugins cc @wargio @thestr4ng3r

File: librz/core/cmd/cmd_api.c [hunk 4]
```diff
@@ -187,7 +185,25 @@ RZ_API void rz_cmd_alias_init(RzCmd *cmd) {
...
+	free(macro->args);
+	free(macro);
+}
+
+static void free_macro_kv(HtPPKv *kv) {
+	free(kv->key);
+	macro_free(kv->value);
+}
+
+RZ_API RzCmd *rz_cmd_new(RzCore *core, bool has_cons) {
```
Comment (XVilka) [inline librz/core/cmd/cmd_api.c:217]: Looks similar to `rz_core_cmd_new()` - maybe noting the differences in the corresponding Doxygen descriptions could help

File: librz/core/cmd/cmd_api.c [hunk 5]
```diff
@@ -1521,491 +1533,234 @@ RZ_API char *rz_cmd_get_help(RzCmd *cmd, RzCmdParsedArgs *args, bool use_color)
...
+/**
+ * \brief Add a new macro named \p name, accepting the arguments \p args, that executes the rizin shell code \p code
+ *
+ * \param cmd Reference to RzCmd
+ * \param name Name of the new macro
+ * \param args Array of strings, terminated by NULL, each string representing the name of the argument
+ * \param code Body of the new macro
+ * \return True when the macro is successfully added, false otherwise
+ */
+RZ_API bool rz_cmd_macro_add(RZ_NONNULL RzCmd *cmd, RZ_NONNULL const char *name, const char **args, const char *code) {
```
Comment (XVilka) [inline librz/core/cmd/cmd_api.c]: Add NONNULL to code as well

File: librz/core/cmd/cmd_api.c [hunk 6]
```diff
@@ -1521,491 +1533,234 @@ RZ_API char *rz_cmd_get_help(RzCmd *cmd, RzCmdParsedArgs *args, bool use_color)
...
+	macro->code = strdup(code);
+	const char **arg = args;
+	for (macro->nargs = 0; *arg; macro->nargs++, arg++) {
 	}
-
-	name = strdup(oname);
-	if (!name) {
-		perror("strdup");
-		return 0;
+	macro->args = RZ_NEWS0(char *, macro->nargs);
```
Comment (XVilka) [inline librz/core/cmd/cmd_api.c:1576]: Check for NULL

File: librz/core/cmd/cmd_api.c [hunk 7]
```diff
@@ -1521,491 +1533,234 @@ RZ_API char *rz_cmd_get_help(RzCmd *cmd, RzCmdParsedArgs *args, bool use_color)
...
+/**
+ * \brief Update an existing macro named \p name, accepting the arguments \p args, that executes the rizin shell code \p code
+ *
+ * \param cmd Reference to RzCmd
+ * \param name Name of the new macro
+ * \param args Array of strings, terminated by NULL, each string representing the name of the argument
+ * \param code Body of the new macro
+ * \return True when the macro is successfully added, false otherwise
+ */
+RZ_API bool rz_cmd_macro_update(RZ_NONNULL RzCmd *cmd, RZ_NONNULL const char *name, const char **args, const char *code) {
```
Comment (XVilka) [inline librz/core/cmd/cmd_api.c]: NONNULL for `code`

File: librz/core/cmd/cmd_api.c [hunk 8]
```diff
@@ -1521,491 +1533,234 @@ RZ_API char *rz_cmd_get_help(RzCmd *cmd, RzCmdParsedArgs *args, bool use_color)
...
+	free(macro->args);
+	for (macro->nargs = 0; args[macro->nargs]; macro->nargs++) {
 	}
-	if (!macro) {
-		macro = rz_cmd_macro_item_new();
-		if (!macro) {
-			free(name);
-			return 0;
-		}
+	macro->args = RZ_NEWS0(char *, macro->nargs);
```
Comment (XVilka) [inline librz/core/cmd/cmd_api.c]: NULL check

File: librz/core/cmd/cmd_api.c [hunk 9]
```diff
@@ -1521,491 +1533,234 @@ RZ_API char *rz_cmd_get_help(RzCmd *cmd, RzCmdParsedArgs *args, bool use_color)
...
-		macro->nargs = rz_str_word_set0(ptr + 1);
-	}
+/**
+ * \brief Remove macro named \p name
+ *
+ * \param cmd Reference to RzCmd
+ * \param name Name of the macro to remove
+ * \return True if the removal was successful, false otherwise
+ */
+RZ_API bool rz_cmd_macro_rm(RZ_NONNULL RzCmd *cmd, const char *name) {
```
Comment (XVilka) [inline librz/core/cmd/cmd_api.c]: Same, NONNULL for `name`

File: librz/core/cmd/cmd_api.c [hunk 10]
```diff
@@ -1521,491 +1533,234 @@ RZ_API char *rz_cmd_get_help(RzCmd *cmd, RzCmdParsedArgs *args, bool use_color)
...
-	free(name);
-	return ret;
+/**
+ * \brief Get a reference to a macro named \p name
+ *
+ * \param cmd Reference to RzCmd
+ * \param name Name of the macro to retrieve
+ * \return Reference to the associated \p RzCmdMacro object
+ */
+RZ_API const RzCmdMacro *rz_cmd_macro_get(RZ_NONNULL RzCmd *cmd, const char *name) {
```
Comment (XVilka) [inline librz/core/cmd/cmd_api.c]: NONULL for `name`

File: librz/core/cmd/cmd_api.c [hunk 11]
```diff
@@ -1521,491 +1533,234 @@ RZ_API char *rz_cmd_get_help(RzCmd *cmd, RzCmdParsedArgs *args, bool use_color)
...
-  )
+/**
+ * \brief Call a macro named \p name passing it the arguments defined in \p argv
+ *
+ * \param cmd Reference to RzCmd
+ * \param name Name of the macro to call
+ * \param argv Array of arguments, terminated by a NULL element
+ * \return Status of the execution of the macros body.
+ */
+RZ_API RzCmdStatus rz_cmd_macro_call(RZ_NONNULL RzCmd *cmd, const char *name, const char **argv) {
```
Comment (XVilka) [inline librz/core/cmd/cmd_api.c]: NONNULL for `name`

File: librz/core/cmd/cmd_api.c [hunk 12]
```diff
@@ -1521,491 +1533,234 @@ RZ_API char *rz_cmd_get_help(RzCmd *cmd, RzCmdParsedArgs *args, bool use_color)
...
+ *
+ * The arguments \p argv should have the number of arguments the macro accepts
+ * (n) or a multiple of it. Each invocation of the macro gets n argument.
+ *
+ * \param cmd Reference to RzCmd
+ * \param name Name of the macro to call
+ * \param argv Array of arguments, terminated by a NULL element
+ * \return Status of the execution of the macros body.
+ */
+RZ_API RzCmdStatus rz_cmd_macro_call_multiple(RZ_NONNULL RzCmd *cmd, const char *name, const char **argv) {
```
Comment (XVilka) [inline librz/core/cmd/cmd_api.c]: NONNULL for `name`

File: librz/core/cmd_descs/cmd_macro.yaml [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    cname: macros_add_call
+    summary: Define a macro <macro-name> and call it with the arguments <macro-call-args>
+    type: RZ_CMD_DESC_TYPE_INNER
+    args_str: "<macro-name> [<macro-arg0> <macro-arg1> ...][; <cmds>])[([<macro-call-arg0> <macro-call-arg1> ...])]"
+  - name: ".("
+    cname: macros_call
+    summary: Call macro <macro-name> with the arguments <macro-call-args>
+    type: RZ_CMD_DESC_TYPE_INNER
+    args_str: "<macro-name> [<macro-call-arg0> <macro-call-arg1> ...])"
+  - name: "..("
```
Comment (XVilka) [inline librz/core/cmd_descs/cmd_macro.yaml:42]: Missed opportunity to use `:(`

**Review by XVilka — CHANGES_REQUESTED** [review]
Overall looks great!

File: librz/core/cmd/cmd.c (hunk 3 above)
Comment (wargio) [reply librz/core/cmd/cmd.c:5237]: not really, this is not called by the plugins.

File: librz/core/cmd_descs/cmd_macro.yaml (hunk 13 above)
Comment (ret2libc) [reply librz/core/cmd_descs/cmd_macro.yaml:42]: :D

**Review by XVilka — APPROVED** [review]
This looks like in good shape and ready to be merged. Does it support nested macroses?

**Comment by ret2libc** [comment]
> This looks like in good shape and ready to be merged. Does it support nested macroses?

```
[0x100003c0c]> (foo a; (_; pd ${a})())
[0x100003c0c]> (
(foo a; (_; pd ${a})())
[0x100003c0c]> .(foo 30)
            ;-- main:
            ;-- entry0:
            ;-- func.100003c0c:
            0x100003c0c      push  rbp
            0x100003c0d      mov   rbp, rsp
            0x100003c10      push  r15
            0x100003c12      push  r14
            0x100003c14      push  r13
            0x100003c16      push  r12
            0x100003c18      push  rbx
            0x100003c19      sub   rsp, 0x638
            0x100003c20      mov   r15, rsi
            0x100003c23      mov   r14d, edi
            0x100003c26      mov   qword [rbp - 0x40], 0
            0x100003c2e      lea   rax, [rbp - 0x260]
            0x100003c35      mov   qword [rbp - 0x48], rax
            0x100003c39      test  edi, edi
        ┌─< 0x100003c3b      jg    0x100003c42
        │   0x100003c3d      call  sym.func.100007237                  ; sym.func.100007237
        └─> 0x100003c42      lea   rsi, section.4.__TEXT.__cstring     ; 0x100007a63
            0x100003c49      xor   edi, edi
            0x100003c4b      call  sym.imp.setlocale                   ; sym.imp.setlocale
            0x100003c50      lea   rdi, str.bin_ls                     ; 0x100007a64 ; "bin/ls"
            0x100003c57      lea   rsi, str.Unix2003                   ; 0x100007a6b ; "Unix2003"
            0x100003c5e      call  sym.imp.compat_mode                 ; sym.imp.compat_mode
            0x100003c63      mov   byte [0x10000856c], al              ; [0x10000856c:1]=0
            0x100003c69      mov   edi, 1
            0x100003c6e      call  sym.imp.isatty                      ; sym.imp.isatty
            0x100003c73      test  eax, eax
        ┌─< 0x100003c75      jne   0x100003cae
        │   0x100003c77      mov   byte section.12.__DATA.__bss, 1     ; [0x100008620:1]=0
        │   0x100003c7e      lea   rdi, str.COLUMNS                    ; 0x100007a74 ; "COLUMNS"
        │   0x100003c85      call  sym.imp.getenv                      ; sym.imp.getenv
[0x100003c0c]> .(foo 10)
            ;-- main:
            ;-- entry0:
            ;-- func.100003c0c:
            0x100003c0c      push  rbp
            0x100003c0d      mov   rbp, rsp
            0x100003c10      push  r15
            0x100003c12      push  r14
            0x100003c14      push  r13
            0x100003c16      push  r12
            0x100003c18      push  rbx
            0x100003c19      sub   rsp, 0x638
            0x100003c20      mov   r15, rsi
            0x100003c23      mov   r14d, edi
```

**Comment by ret2libc** [comment]
> This looks like in good shape and ready to be merged. Does it support nested macroses?

I want to reorganize a bit the commits before merging.

**Comment by ret2libc** [comment]
Unfortunately the commit `@ret2libc
[core/cmd: rewrite rz_cmd_macro API](https://github.com/rizinorg/rizin/pull/2915/commits/de40ccc77c9d05362b13650691a44ae7d1a454ce)`  is quite big, as the changes in the grammar require a lot of changes in the codebase.

I added unit tests and there are tests for macros anyway. I should also fix the book.

**Review by XVilka — APPROVED** [review]

**Pull request merged**